### PR TITLE
feat: [CDS-85968]: add callback id handling for approval step and api

### PR DESCRIPTION
**Pull request by RishabhGupta34**
## Describe your changes
Added callback Id in the harness approval step params, this is another id which can be used by the customer to identify an approval instance and approve it.
Have modified the API for approving by execution id to have callback id as a parameter.

Tested locally the changes:
1. Passing callback id to the API and approving/rejecting an execution.
2. Passing no callback id to the API and approving/rejecting an execution.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56404)

### CODE REVIEW

**Review by mankrit-singh — DISMISSED** [review]

**Comment by RishabhGupta34** [comment]
trigger ut5, ut6, ut9

**Comment by RishabhGupta34** [comment]
trigger ut5, ut9

**Review by puthrayaharness — DISMISSED** [review]

**Comment by RishabhGupta34** [comment]
trigger review

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/harness/entities/HarnessApprovalInstance.java [hunk 1]
```diff
@@ -155,8 +156,13 @@ public static HarnessApprovalInstance fromStepParameters(Ambiance ambiance, Step
...
     }
     HarnessApprovalSpecParameters specParameters = (HarnessApprovalSpecParameters) stepParameters.getSpec();
+    String callbackId = null;
+    if (specParameters.getCallbackId() != null) {
+      callbackId = (String) specParameters.getCallbackId().fetchFinalValue();
+    }
     HarnessApprovalInstance instance =
         HarnessApprovalInstance.builder()
+            .callbackId(callbackId)
             .approvalMessage((String) specParameters.getApprovalMessage().fetchFinalValue())
```
Comment (naman-goenka) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/harness/entities/HarnessApprovalInstance.java:166]: please add index for callback id in approval instance

File: pipeline-service/contracts/openapi/v1/openapi.yaml [hunk 2]
```diff
@@ -1067,15 +1072,15 @@ paths:
...
+        id: wnd53l2n474n1
       description: Retry a executed Pipeline with inputSet Pipeline yaml.
       parameters:
         - $ref: '#/components/parameters/AccountHeader'
         - schema:
             type: string
           in: query
           name: module
-          description: Module type for the entity. If its from deployments,type will be CD , if its from build type will be CI.
+          description: 'Module type for the entity. If its from deployments,type will be CD , if its from build type will be CI.'
```
Comment (naman-goenka) [inline pipeline-service/contracts/openapi/v1/openapi.yaml:1083]: in this an intended change

File: pipeline-service/service/src/main/java/io/harness/pms/approval/api/ApprovalsApiImpl.java [hunk 3]
```diff
@@ -129,7 +129,7 @@ public Response getApprovalInstancesByExecutionId(@OrgIdentifier String org, @Pr
     }
 
     List<ApprovalInstanceResponseDTO> approvalInstances = approvalResourceService.getApprovalInstancesByExecutionId(
-        executionId, approvalStatusEnum, approvalTypeEnum, nodeExecutionId);
+        executionId, approvalStatusEnum, approvalTypeEnum, nodeExecutionId, null);
```
[the remark this answers is not in the document]
Comment (RishabhGupta34) [reply pipeline-service/service/src/main/java/io/harness/pms/approval/api/ApprovalsApiImpl.java:132]: will create a follow up ticket to support it in this api as well

File: pipeline-service/service/src/main/java/io/harness/pms/approval/ApprovalResourceServiceImpl.java [hunk 4]
```diff
@@ -134,9 +134,9 @@ public ApprovalInstanceResponseDTO addHarnessApprovalActivity(
...
       @NotNull @ProjectIdentifier String projectIdentifier, @NotNull String planExecutionId,
-      @NotNull @Valid HarnessApprovalActivityRequestDTO request) {
-    List<ApprovalInstanceResponseDTO> approvalInstances =
-        getApprovalInstancesByExecutionId(planExecutionId, ApprovalStatus.WAITING, ApprovalType.HARNESS_APPROVAL, null);
+      @NotNull @Valid HarnessApprovalActivityRequestDTO request, String callbackId) {
+    List<ApprovalInstanceResponseDTO> approvalInstances = getApprovalInstancesByExecutionId(
+        planExecutionId, ApprovalStatus.WAITING, ApprovalType.HARNESS_APPROVAL, null, callbackId);
     if (EmptyPredicate.isEmpty(approvalInstances)) {
       instrumentationHelper.sendApprovalApiEvent(accountId, orgIdentifier, projectIdentifier, planExecutionId,
           ApprovalApiInstrumentationHelper.FAILURE, ApprovalApiInstrumentationHelper.NO_APPROVALS_FOUND);
```
[the remark this answers is not in the document]
Comment (RishabhGupta34) [reply pipeline-service/service/src/main/java/io/harness/pms/approval/ApprovalResourceServiceImpl.java:142]: added

File: pipeline-service/contracts/openapi/v1/openapi.yaml (hunk 2 above)
Comment (RishabhGupta34) [reply pipeline-service/contracts/openapi/v1/openapi.yaml]: nice catch, fixed

**Comment by RishabhGupta34** [comment]
trigger uts

**Comment by RishabhGupta34** [comment]
trigger ut5, ut7

**Comment by RishabhGupta34** [comment]
trigger ut7

**Comment by RishabhGupta34** [comment]
trigger ut7

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/harness/entities/HarnessApprovalInstance.java [hunk 5]
```diff
@@ -174,6 +175,9 @@ public static HarnessApprovalInstance fromStepParameters(Ambiance ambiance, Step
             .approvalKey(stageIdentifier + "#" + stepParameters.getIdentifier())
             .autoApproval(fromAutoApprovalParams(specParameters.getAutoApproval()))
             .build();
+    if (specParameters.getCallbackId() != null) {
```
[the remark this answers is not in the document]
Comment (RishabhGupta34) [reply pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/harness/entities/HarnessApprovalInstance.java:178]: that gives the same response, as the not null check

**Comment by abhinavmittal-harness** [comment]
trigger review

**Pull request merged**